lineapy/plugins/kubeflow_pipeline_writer.py

```diff
@@ -0,0 +1,166 @@
+import logging
+from enum import Enum
+from typing import Any, Dict, List, Tuple
+
+from typing_extensions import TypedDict
+
+from lineapy.plugins.base_pipeline_writer import BasePipelineWriter
+from lineapy.plugins.task import (
+    DagTaskBreakdown,
+    TaskDefinition,
+    TaskSerializer,
+    render_task_io_serialize_blocks,
+)
+from lineapy.plugins.taskgen import get_task_graph
+from lineapy.plugins.utils import load_plugin_template
+from lineapy.utils.logging_config import configure_logging
+from lineapy.utils.utils import prettify
+
+logger = logging.getLogger(__name__)
+configure_logging()
+
+
+class KubeflowDagFlavor(Enum):
+    ComponentPerSession = 1
+    ComponentPerArtifact = 2
+
+
+KubeflowDagConfig = TypedDict(
+    "KubeflowDagConfig",
+    {
+        "host_url": str,
+        "dag_flavor": str,  # Not native to DVC config
+    },
+    total=False,
+)
+
+
+class KubeflowPipelineWriter(BasePipelineWriter):
+    def _write_dag(self) -> None:
+        # Check if the given DAG flavor is a supported/valid one
+        try:
+            dag_flavor = KubeflowDagFlavor[
+                self.dag_config.get("dag_flavor", "ComponentPerArtifact")
+            ]
+        except KeyError:
+            raise ValueError(
+                f'"{dag_flavor}" is an invalid kubeflow dag flavor.'
+            )
+
+        # Construct DAG text for the given flavor
+        full_code = self._write_operators(dag_flavor)
+
+        # Write out file
+        file = self.output_dir / f"{self.pipeline_name}_dag.py"
+        file.write_text(full_code)
+        logger.info(f"Generated DAG file: {file}")
+
+    def _write_operators(
+        self,
+        dag_flavor: KubeflowDagFlavor,
+    ) -> str:
+        """
+        Returns a code block containing all the operators for a Kubeflow DAG.
+        """
+
+        DAG_TEMPLATE = load_plugin_template("kubeflow_dag.jinja")
+
+        if dag_flavor == KubeflowDagFlavor.ComponentPerSession:
+            task_breakdown = DagTaskBreakdown.TaskPerSession
+        elif dag_flavor == KubeflowDagFlavor.ComponentPerArtifact:
+            task_breakdown = DagTaskBreakdown.TaskPerArtifact
+
+        # Get task definitions based on dag_flavor
+        task_defs, _ = get_task_graph(
+            self.artifact_collection,
+            pipeline_name=self.pipeline_name,
+            task_breakdown=task_breakdown,
+        )
+
+        task_names = list(task_defs.keys())
+
+        task_defs = {tn: task_defs[tn] for tn in task_names}
+
```

Comment:
add the setup and teardown tasks to keep the hooks that we provide our users (this is true of every type of pipeline). In the current case, it will only contain pass since we arent setting anything up but if the user wants to ensure that latest data gets copied or to save the final result of the pipeline to some place, they can use these hooks and add their custom code. Also not a show stopper but a really nice thing to have.

Reply:
Added task teardown and cleanup
![image](https://user-images.githubusercontent.com/7387520/208525749-bb5450f2-5606-4af8-bf14-fd44d75b7de5.png)